File: packages/python/qsu/object/objDeleteKeyByValue.py
from ..verify.isObject import isObject


def objDeleteKeyByValue(obj, searchValue, recursive=False):
	if not isinstance(obj, dict):
		return None

	# Work on a copy. `newObj = obj` modified the caller's dict in place.
	newObj = dict(obj)

	# Start at the last index. Starting at `len(keys)` read one index past the end, which
	# is why the guard below existed at all.
	keys = list(newObj.keys())

	for i in range(len(keys) - 1, -1, -1):
		key = keys[i]

		if recursive and newObj.get(key) is not None and isObject(newObj[key]):
			newObj[key] = objDeleteKeyByValue(newObj[key], searchValue, recursive)
		elif _strictEqual(newObj.get(key), searchValue):
			del newObj[key]

	return newObj


def _strictEqual(a, b) -> bool:
	if isinstance(a, bool) != isinstance(b, bool):
		return False
	return a == b

File: packages/javascript/lib/object/objDeleteKeyByValue.ts
import type { AnyValueObject } from '../_types/global';
import { isObject } from '../verify/isObject.js';

export function objDeleteKeyByValue(
	obj: AnyValueObject,
	searchValue: string | number | null | undefined,
	recursive = false
): AnyValueObject | null {
	if (!obj || typeof obj !== 'object') {
		return null;
	}

	// Work on a copy. `Object.assign(obj, {})` returned `obj` itself, so the caller's object
	// was modified in place.
	const newObj: AnyValueObject = { ...obj };

	// Build the key list once. Calling `Object.keys` inside the loop rebuilt the whole array
	// on every iteration, and starting at `length` read one index past the end.
	const keys = Object.keys(newObj);

	for (let i = keys.length - 1; i >= 0; i -= 1) {
		const key = keys[i];

		if (recursive && newObj[key] && isObject(newObj[key])) {
			newObj[key] = objDeleteKeyByValue(newObj[key], searchValue, recursive);
		} else if (newObj[key] === searchValue) {
			delete newObj[key];
		}
	}

	return newObj;
}
